sshpiperd/upstream/database/handler.go

```diff
@@ -2,86 +2,313 @@ package database
+				}
+
+				return ssh.AuthPipeTypePassThrough, nil, nil
+			}
+		}
 
+	case authMapTypePrivateKey:
+
+		allowAnyPubKey = allowAnyPubKey || pipe.FromAllowAnyPublicKey
+
+		if !allowAnyPubKey {
+			for _, privkey := range pipe.FromAuthorizedKeys {
+				rest, err := p.GetBytes(privkey.Key.Data)
 				if err != nil {
-					logger.Printf("parse [keyid = %v] error :%v. skip to next key", k.Key.ID, err)
-					continue
+					return nil, err
 				}
 
-				if bytes.Equal(publicKey.Marshal(), expectKey) {
+				var authedPubkey ssh.PublicKey
 
-					kinterf, err := ssh.ParseRawPrivateKey([]byte(d.Upstream.PrivateKey.Key.Data))
+				for len(rest) > 0 {
+					authedPubkey, _, _, rest, err = ssh.ParseAuthorizedKey(rest)
 					if err != nil {
-						break
+						return nil, err
 					}
 
-					signer, err := ssh.NewSignerFromKey(kinterf)
-					if err != nil || signer == nil {
-						break
+					allowPubKeys = append(allowPubKeys, authedPubkey)
+				}
+			}
+		}
+
+		if a.PublicKeyCallback == nil {
+			a.PublicKeyCallback = func(conn ssh.ConnMetadata, key ssh.PublicKey) (ssh.AuthPipeType, ssh.AuthMethod, error) {
+
+				if allowAnyPubKey {
+					return to(key)
+				}
+
+				keydata := key.Marshal()
+
+				for _, authedPubkey := range allowPubKeys {
+					if bytes.Equal(authedPubkey.Marshal(), keydata) {
+						return to(key)
 					}
+				}
 
-					return ssh.AuthPipeTypeMap, ssh.PublicKeys(signer), nil
+				if pipe.NoPassthrough {
+					return ssh.AuthPipeTypeDiscard, nil, nil
 				}
+
+				// will fail but discard will lead a timeout
+				return ssh.AuthPipeTypePassThrough, nil, nil
 			}
+		}
 
-			return ssh.AuthPipeTypeNone, nil, nil
-		},
+	case authMapTypeAny:
+		a.NoneAuthCallback = func(conn ssh.ConnMetadata) (ssh.AuthPipeType, ssh.AuthMethod, error) {
+			return to(nil)
+		}
 
-		UpstreamHostKeyCallback: hostKeyCallback,
+		a.PasswordCallback = func(conn ssh.ConnMetadata, password []byte) (ssh.AuthPipeType, ssh.AuthMethod, error) {
+			return to(nil)
+		}
+
+		a.PublicKeyCallback = func(conn ssh.ConnMetadata, key ssh.PublicKey) (ssh.AuthPipeType, ssh.AuthMethod, error) {
+			return to(key)
+		}
+
+		return a, nil
+
+	default:
+		logger.Printf("unsupport type [%v], ignore section", pipe.FromType)
+	}
+
+	return a, nil
+
+}
+
+func (p *plugin) findUpstream(conn ssh.ConnMetadata, challengeContext ssh.AdditionalChallengeContext) (net.Conn, *ssh.AuthPipe, error) {
+
+	pipe, err := p.loadPipeFromDB(conn)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if pipe.Username != "" {
+
+		logger.Printf("mapping [%v] to [%v@%v] from %v to %v", pipe.Username, pipe.MappedUsername, pipe.UpstreamHost, pipe.FromType, pipe.ToType)
+
+		c, err := upstreamprovider.DialForSSH(pipe.UpstreamHost)
+		if err != nil {
+			return nil, nil, err
+		}
+		a, err := p.createAuthPipe(pipe, conn, challengeContext)
+		if err != nil {
+			return nil, nil, err
+		}
+		return c, a, nil
 	}
-	return c, &pipe, nil
+
+	return nil, nil, fmt.Errorf("username not [%v] found", pipe.Username)
```

Comment:
empty "" not found? or it cannot be empty?

Reply:
should not be empty, modified